// rebylon/dist/immediate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isPromise(value) {
    return value !== null && typeof value === "object" && typeof value.then === "function";
}
class ResolvedImmediate {
    constructor(value) {
        this.value = value;
    }
    then(onfulfilled, _) {
        if (onfulfilled) {
            const value = onfulfilled(this.value);
            if (isPromise(value)) {
                return new Immediate(value);
            }
            else {
                return new ResolvedImmediate(value);
            }
        }
        else {
            return this;
        }
    }
}
class RejectedImmediate {
    constructor(reason) {
        this.reason = reason;
    }
    then(_, onrejected) {
        if (onrejected) {
            const value = onrejected(this.reason);
            if (isPromise(value)) {
                return new Immediate(value);
            }
            else {
                return new ResolvedImmediate(value);
            }
        }
        else {
            return this;
        }
    }
}
class Immediate {
    constructor(promise) {
        this.promise = promise;
        this.result = null;
        if (promise instanceof ResolvedImmediate || promise instanceof RejectedImmediate || promise instanceof Immediate) {
            this.result = promise;
        }
        else {
            promise.then(value => {
                this.result = new ResolvedImmediate(value);
            }, reason => {
                this.result = new RejectedImmediate(reason);
            });
        }
    }
    then(onfulfilled, onrejected) {
        if (this.result) {
            return this.result.then(onfulfilled, onrejected);
        }
        else {
            return this.promise.then(onfulfilled, onrejected);
        }
    }
    static resolve(value) {
        return new ResolvedImmediate(value);
    }
    static reject(reason) {
        return new RejectedImmediate(reason);
    }
    static wrap(callback) {
        return (...a) => new Immediate(callback(...a));
    }
}
exports.Immediate = Immediate;
